### PR TITLE
[Ready for Review] Feature/addon error handling

### PR DESCRIPTION
**Pull request by samchrisinger**
## Purpose

QA noticed some places where we provide little or no feedback to the user when credentials are invalid or expired. https://github.com/CenterForOpenScience/osf.io/issues/2495#issuecomment-93752701

## Changes
- Add error handling to addon Node config
- Improve S3 error handling and UI

## Side Effects 
None

### CODE REVIEW

File: website/addons/s3/model.py [hunk 1]
```diff
@@ -56,6 +57,10 @@ def to_json(self, user):
     def has_auth(self):
         return bool(self.access_key and self.secret_key)
 
+    @property
+    def is_valid(self):
+        return api.has_access(self.access_key, self.secret_key)
```
Comment (sloria) [inline website/addons/s3/model.py]: Would it make sense to make this a cached property?

File: website/addons/s3/utils.py [hunk 2]
```diff
@@ -48,9 +49,11 @@ def get_bucket_drop_down(user_settings):
         return [
             bucket.name
             for bucket in get_bucket_list(user_settings)
-        ]
-    except BotoServerError:
-        return None
+        ], http.OK
+    except BotoServerError as e:
+        return None, e.status
```
Comment (sloria) [inline website/addons/s3/utils.py]: Why not raise an HTTPError instead of returning a tuple?

File: website/addons/s3/utils.py [hunk 3]
```diff
@@ -48,9 +49,11 @@ def get_bucket_drop_down(user_settings):
...
+    except NoAuthHandlerFound:
+        return None, http.FORBIDDEN
```
Comment (sloria) [inline website/addons/s3/utils.py]: Same here: raise an `HTTPError`?

File: website/addons/s3/views/config.py [hunk 4]
```diff
@@ -149,10 +149,13 @@ def s3_get_node_settings(auth, node_addon, **kwargs):
...
-
-    return {
-        'buckets': get_bucket_drop_down(user_addon)
-    }
+    buckets, status_code = get_bucket_drop_down(user_addon)
+    if buckets:
+        return {
+            'buckets': buckets
+        }
+    else:
```
Comment (sloria) [inline website/addons/s3/views/config.py]: If you raise an `HTTPError` in `get_bucket_drop_down` as suggested in my previous comment, this branch is unnecessary.

**Comment by sloria** [comment]
Pass finished. :point_up:

File: website/addons/s3/model.py (hunk 1 above)
Comment (samchrisinger) [inline website/addons/s3/model.py]: I don't think caching makes sense here. The goal of adding this property is to detect expired and invalid tokens before a user tries to get a list of their S3 buckets. 

Alternatively we can wait until they try to fetch buckets, and then propagate the error from the S3 UI.

File: website/static/js/folderPickerNodeConfig.js [hunk 5]
```diff
@@ -209,11 +209,16 @@ var FolderPickerViewModel = oop.defclass({
         this.messageClass('text-info');
     },
     /**
+     * Abstract hook called after updateFromData, before the promise is resolved.
+     * - use to validate the VM state after update     
+     **/
+    _afterUpdateHook: function() {},
```
Comment (sloria) [inline website/static/js/folderPickerNodeConfig.js]: For brevity and consistency, how about just `afterUpdate` or `postUpdate`?

**Comment by sloria** [comment]
Pass finished. :v: Just a minor naming issue, and this is good to merge.

**Pull request merged**